pandas/core/nanops.py

```diff
@@ -498,18 +501,45 @@ def nansum(
     >>> nanops.nansum(s)
     3.0
     """
+    orig_values = values
+
     values, mask, dtype, dtype_max, _ = _get_values(
         values, skipna, fill_value=0, mask=mask
     )
     dtype_sum = dtype_max
+    datetimelike = False
     if is_float_dtype(dtype):
         dtype_sum = dtype
     elif is_timedelta64_dtype(dtype):
+        datetimelike = True
         dtype_sum = np.float64
+        if mask is None and not skipna:
+            mask = isna(orig_values)
+
     the_sum = values.sum(axis, dtype=dtype_sum)
     the_sum = _maybe_null_out(the_sum, axis, mask, values.shape, min_count=min_count)
 
-    return _wrap_results(the_sum, dtype)
+    the_sum = _wrap_results(the_sum, dtype)
+    if datetimelike and not skipna:
+        the_sum = _mask_datetimelike_result(the_sum, axis, mask, orig_values.dtype)
+    return the_sum
+
+
+def _mask_datetimelike_result(
```

Comment:
why don;t you have this take the original values (rather than just the dtype) and compute the mask if needed (e.g. make it optional), rn the caller is responsible for that in multiple places.

Reply:
if we get to here, we always need the mask

Reply:
you didn't answer the question. you are adding multiple code blocks which do the same thing; if you are going to consolidate to a function then it makes sense to avoid that yes?

Reply:
apparently i dont understand the question.  IIUC the alternative you're suggesting looks like

```
def _mask_datetimelike_result(result, axis, mask, orig_values):
    if mask is None:
        mask = isna(orig_values)
    [what we have here now]
```

and remove the `if mask is None and not skipna: mask = isna(orig_values)` on L516-517.  This is 2 lines of code either way, so not a big deal.  ill change it if you really care.

longer-term this should probably go into `_get_values`, but i want to do that carefully since that may affect other functions

Reply:
ok i guess - i am thinking u r going to refactor this anyhow as this adds a fair amount of duplication